Fix backlink check in migration.

DominicFrei opened the pull request:
Necessary core changes to enable https://github.com/realm/realm-cocoa/pull/7094 (https://github.com/realm/realm-cocoa/issues/7060):
- Setting tables to embedded had to be moved to post migration changes.
- Objects without backlinks will get deleted.

fealebenpae commented:
Hey @DominicFrei, thanks for the contribution! There are just a couple of things we need to fix: please update `CHANGELOG.md` with a line explaining the change (see `CONTRIBUTING.md`), and you need to run `clang-format` on your patch to make sure it's formatted according to the repository rules (see the [Artifacts](https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-core/detail/PR-4414/1/artifacts) tab in CI for the format errors).
@jedelbo could you please review this?

DominicFrei commented:
@fealebenpae Thanks for clearing that up! I've put it into draft mode since there seems to be one thing to clarify first. @tgoyne suggested a couple of tests that I implemented. According to the CI, one (already existing) test is failing though:
`TEST(Group_ChangeEmbeddedness)` 
It checks if all objects have a parent when their table is set to `embedded`.
Thomas' suggestion for one of the new tests was:
`Objects with zero incoming links are automatically deleted when converting to embedded`
Those two tests conflict now if I understand it correctly, and I first need to understand which of the two is the desired behaviour.

jedelbo commented on src/realm/table.cpp:
```diff
@@ -1053,12 +1053,20 @@ bool Table::set_embedded(bool embedded)
             }
+        } else {
+            // If the table does not have any backlink columns the objects can be deleted.
+            for (auto object : *this) {
```
If the table has no backlink columns, then the objects cannot be owned by objects in another table. So the table cannot be embedded.
Anyway - I would prefer `clear()` here.

ironage replied:
Having an embedded type with no incoming links will cause a schema validation error in AdditiveExplicit mode (since https://github.com/realm/realm-core/pull/4306). There is clearly some error here but I'm not sure if we should allow it here and fail later or what. To me it seems likely that the developer made a mistake, and if the change to embedded was an accident or test, it might be unwelcome to have all the data erased. For that reason, I suggest throwing an error here. Perhaps the error message can tell the user to first make a link to the table before changing it to be embedded.

jedelbo commented on src/realm/table.cpp:
```diff
@@ -1053,12 +1053,20 @@ bool Table::set_embedded(bool embedded)
+            for (auto object : *this) {
+                if (object.get_backlink_count() == 0) {
+                    // If the object does not have any backlinks it can be deleted.
+                    object.remove();
```
Not sure it is a good idea to remove objects if the whole table does not fulfill the criteria of being embedded.

ironage commented on src/realm/table.cpp:
```diff
@@ -1053,13 +1053,23 @@ bool Table::set_embedded(bool embedded)
         });
 
```
Shouldn't we skip all the following checks when migrating from embedded to non-embedded (eg: `set_embedded(false)`)?

DominicFrei commented:
According to the [original post in the forum](https://developer.mongodb.com/community/forums/t/migrating-to-using-the-new-embedded-object-in-mongodb-realm/11920) that brought this problem up it seems like that specific user did actually have `child objects` being used by multiple `parent objects`. This did fail in the past and will still fail after this PR but the user will have a chance to fix it within the migration block from now on due to calling `set_embedded` in the `apply_post_migration_changes` instead of the `apply_pre_migration_changes` (see `src/realm/object-store/object_store.cpp`).

Maybe the user was mislead by the error message in https://github.com/realm/realm-core/blob/master/src/realm/object-store/object_store.cpp#L570 though:
`"Cannot convert object type '%1' to embedded because objects have multiple incoming links."`
This error is thrown whenever `set_embedded` returns `false` which can happen in multiple cases, not just `if (o.get_backlink_count() != 1)`.

To address this issue a bit better and also incorporate the suggestions made by @jedelbo and @ironage I reworked the `set_embedded` function:
- Move the above mentioned error into `set_embedded` and only throw this particular one in case there are multiple incoming links.
- Add new `throws` in all other cases where we returned `false` for reasons other than multiple backlinks.
- Remove suggested deletion of objects and throw an error in all those cases.
- Add a check for `isEmbedded == false` as suggested by @ironage to skip unnecessary checks.

Notes:
- I also tried to add a unit test for the `if (get_primary_key_column())` path (since I reworked the whole function anyway). The test passes but a different reason (tbd).
- The path `if (embedded == m_is_embedded)` is actually never hit since https://github.com/realm/realm-core/blob/master/src/realm/object-store/shared_realm.cpp#L344 evaluates the `required_changes` to be empty which means `set_embedded` is never called anyway.

ironage commented on test/object-store/migrations.cpp:
```diff
@@ -577,29 +577,267 @@ TEST_CASE("migration: Automatic") {
             realm->commit_transaction();
+            REQUIRE(parent_table->size() == 2);
+            REQUIRE(child_table->size() == 2);
+
```
suggest adding some checks around `child_table->is_embedded()` before and after the migration

ironage commented on test/object-store/migrations.cpp:
```diff
@@ -577,29 +577,267 @@ TEST_CASE("migration: Automatic") {
+            }
+        }
+        
+        SECTION("change table to embedded - adding more links in migration block") {
```
really great to see all these cases tested 💯

DominicFrei replied:
Thanks. :)

ironage reviewed (approved):
👍

sipersso commented:
> According to the [original post in the forum](https://developer.mongodb.com/community/forums/t/migrating-to-using-the-new-embedded-object-in-mongodb-realm/11920) that brought this problem up it seems like that specific user did actually have `child objects` being used by multiple `parent objects`. This did fail in the past and will still fail after this PR but the user will have a chance to fix it within the migration block from now on due to calling `set_embedded` in the `apply_post_migration_changes` instead of the `apply_pre_migration_changes` (see `src/realm/object-store/object_store.cpp`).
> 
> Maybe the user was mislead by the error message in https://github.com/realm/realm-core/blob/master/src/realm/object-store/object_store.cpp#L570 though:
> `"Cannot convert object type '%1' to embedded because objects have multiple incoming links."`
> This error is thrown whenever `set_embedded` returns `false` which can happen in multiple cases, not just `if (o.get_backlink_count() != 1)`.
> 
> To address this issue a bit better and also incorporate the suggestions made by @jedelbo and @ironage I reworked the `set_embedded` function:
> 
> * Move the above mentioned error into `set_embedded` and only throw this particular one in case there are multiple incoming links.
> * Add new `throws` in all other cases where we returned `false` for reasons other than multiple backlinks.
> * Remove suggested deletion of objects and throw an error in all those cases.
> * Add a check for `isEmbedded == false` as suggested by @ironage to skip unnecessary checks.
> 
> Notes:
> 
> * I also tried to add a unit test for the `if (get_primary_key_column())` path (since I reworked the whole function anyway). The test passes but a different reason (tbd).
> * The path `if (embedded == m_is_embedded)` is actually never hit since https://github.com/realm/realm-core/blob/master/src/realm/object-store/shared_realm.cpp#L344 evaluates the `required_changes` to be empty which means `set_embedded` is never called anyway.

I am "the user" from the forums. Just to clarify, my objects that I wanted to embed did not have multiple parents even though the crash said that they did, so the error message was misleading. 

But... one thing I am scared about is orphaned objects. I would not be able to use this in production without a way to ensure that I have deleted all possible orphaned objects. Crashes in the migration function are about as bad as it can get from a user perspective.

DominicFrei commented:
@sipersso The changes made in https://github.com/realm/realm-core/pull/4414/files#diff-e752b38a6d72eb50c7ab147fe48749dee382b2c4153c01364db035c2843f1519R1054 lead to several different errors that will be thrown depending on the problem that happened. The error message should be accurate now.
